pkg/release/build_version.go

```diff
@@ -0,0 +1,391 @@
+
+	mainJobCache, err := b.jobCacheClient.GetJobCache(mainJob, true)
+	if err != nil {
+		return "", errors.Wrap(err, "building job cache for main job")
+	}
+	if mainJobCache == nil {
+		return "", errors.Errorf("main job cache for job %q is nil", mainJob)
+	}
+
+	// Update secondary caches limited by main cache last build number
+	secondaryJobCaches := []*JobCache{}
+	for _, job := range secondaryJobs {
+		cache, err := b.jobCacheClient.GetJobCache(job, true)
+		if err != nil {
+			return "", errors.Wrapf(err, "building job cache for job: %s", job)
+		}
+		if cache != nil {
+			secondaryJobCaches = append(secondaryJobCaches, cache)
+		}
+	}
+
+	for i, version := range mainJobCache.Versions {
+		sb := strings.Builder{}
+		tw := tabwriter.NewWriter(&sb, 0, 0, 2, ' ', 0)
+		fmt.Fprintln(tw, "Job\tRun\tBuild\tDate/Status")
+
+		buildVersion := mainJobCache.BuildNumbers[i]
+
+		logrus.Infof("Trying version %q for build %q", version, buildVersion)
+
+		if i > jobLimit {
+			return "", errors.Errorf("job count limit of %d exceeded", jobLimit)
+		}
+
+		matches := regex.ReleaseAndBuildRegex.FindStringSubmatch(version)
+		if matches == nil || len(matches) < 8 {
+			return "", errors.Errorf("invalid build version: %v", version)
+		}
+
+		buildRun := matches[6]
+		buildSHA := matches[7]
+
+		date, err := b.githubClient.GetCommitDate(buildSHA)
+		if err != nil {
+			return "", errors.Wrapf(err, "retrieve repository commit %s", buildSHA)
+		}
+
+		fmt.Fprint(&sb, "(*) Primary job (-) Secondary jobs\n\n")
+		fmt.Fprintf(tw,
+			"%s\t%s\t%s\t%s\t\n",
+			"* "+strings.TrimPrefix(mainJob, jobPrefix),
+			buildVersion, buildRun, date,
+		)
+
+		type BuildStatus string
+		const (
+			BuildStatusNotExisting BuildStatus = "NOT EXISTING"
+			BuildStatusFailed      BuildStatus = "FAILED"
+			BuildStatusSucceeded   BuildStatus = "SUCCEEDED"
+		)
+		// Check secondaries to ensure that build number is green across all
+		success := true
+		foundBuildNumber := ""
+		for _, secondaryJob := range secondaryJobs {
+			status := BuildStatusNotExisting
+			for _, secondaryJobCache := range secondaryJobCaches {
+				if secondaryJobCache.Name == secondaryJob {
+					status = BuildStatusFailed
+
+					for j, secVersion := range secondaryJobCache.Versions {
+						matches := regex.ReleaseAndBuildRegex.FindStringSubmatch(secVersion)
+						if matches == nil || len(matches) < 8 {
+							logrus.Errorf(
+								"Invalid build version %s for job %s",
+								secVersion, secondaryJob,
+							)
+							continue
+						}
+
+						// Verify that we have the same build number
+						if buildRun == matches[6] {
+							status = BuildStatusSucceeded
+							foundBuildNumber = secondaryJobCache.BuildNumbers[j]
+						}
+					}
+					if status == BuildStatusSucceeded {
+						break
+					}
+				}
+			}
+
+			fmt.Fprintf(tw,
+				"%s\t%s\t%s\t%s\t\n",
+				"- "+strings.TrimPrefix(secondaryJob, jobPrefix),
+				foundBuildNumber, buildRun, status,
+			)
+
+			if status == BuildStatusFailed {
```

Comment:
Are folks comfortable with the behavior when `status == BuildStatusNotExisting` implying `success == true` and GREEN?  This worries me.

Reply:
@kubernetes/release-managers and @kubernetes/ci-signal what do you think of this as a health metric?

Reply:
when would state be `BuildStatusNotExisting`? if it only happens when the build doesn't finish yet, then I think it is ok.

Reply:
Running the tool I see it showing quite commonly.  If say half the tests on a given build have no existing status, the other half exist and are passing, does that constitute a successfully passing build?

Reply:
Yeah I think that jobs like `integration-master` do not exists is related to the deprecation of the job build cache. :-/

Reply:
I should have saved a screen shot.  The other day I could swear I had one of
```
conformance-kind-ga-only
integration-master
kind-e2e-parallel
kind-ipv6-e2e-parallel
verify-master
periodic-kubernetes-bazel-build-master
periodic-kubernetes-bazel-test-master
```
show as failed or success instead of just not existing.

If those can fail, and if a failure would be blocking, then we need a way around the "not existing" status to determine pass or fail.  Otherwise the tool could lead to poor choices.

Given the current minimal scope imagined for the tool, perhaps for tests could be ignored.  We just need to do that knowingly, not do it in a critical decision path, and have an action logged to resolve this.